### DISCUSSION TITLE
Transfer learning with frozen layers - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#3276] [Docs]
# Transfer learning with frozen layers - Ultralytics YOLOv8 Docs

Learn how to freeze YOLOv5 when transfer learning. Retrain a pre-trained model on new data faster and with fewer resources.

https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/

**Comment by 4yafes6** [comment]
Thank you for your informative response, but there is a point that I didn't understand and would like to ask about. In the yolov5m.pt file, there are 80 classes, but I have class 'A'. I have organized the images and their corresponding labels for class 'A' into train, validation, and test folders. I want to add class 'A' to the yolov5m.pt file, making it a total of 81 classes. However, I only want to train class 'A' and not retrain the other 80 classes. What do I need to do for this?

**Reply by glenn-jocher** [reply]
@4yafes6 to add a new class 'A' to an existing YOLOv8 model pre-trained on 80 classes without retraining the other classes, you'll need to perform transfer learning with layer freezing. Here's a high-level overview of the steps you should follow:

1. **Update your dataset**: Ensure your dataset is properly labeled for class 'A', with images and annotations in the correct format.

2. **Modify the model configuration**: Adjust the number of classes in your model configuration file to 81 to account for the new class 'A'.

3. **Freeze layers**: When setting up your training configuration, specify which layers of the model you want to freeze. Typically, you would freeze the backbone and only train the head of the network where the class predictions are made.

4. **Train the model**: Start the training process with the modified configuration, ensuring that only the layers responsible for class 'A' are being updated.

Remember, when you freeze layers, the pre-trained weights for the original 80 classes remain unchanged, and only the weights for the new class 'A' are learned. This approach is efficient and leverages the existing knowledge of the model while expanding its capabilities to recognize an additional class.

For detailed instructions and the specific commands to use, please refer to the Ultralytics YOLOv8 documentation. 😊

**Comment by Rayen023** [comment]
Would like to understand how is the model still learning with all layers frozen ?

**Reply by nevercast** [reply]
All **except** the final detect layer.

**Reply by glenn-jocher** [reply]
@nevercast when you freeze layers during transfer learning, you're essentially telling the model to keep the weights of those layers constant. This means that the frozen layers will not update during the training process. However, not all layers are frozen; typically, only the early layers responsible for capturing basic features are frozen.

The final layers, such as the detection layer in YOLOv8, remain unfrozen and continue to learn. These layers are more specialized towards the task at hand, such as detecting specific objects in a new dataset. By keeping the early layers frozen, the model can leverage the general features learned from the original dataset, while the unfrozen layers adapt to the new data, allowing for efficient transfer learning. This approach can lead to faster training times and reduced computational resources, as only a portion of the model's parameters are being updated. 🧊➡️🔥🎓

**Reply by nevercast** [reply]
Thank you for the detailed reply. Certainly, clearer than my response though getting at the same thing. Not all layers are frozen, the final detection layers are unfrozen, which was my intent to convey with `except the final detect layer`.  I appreciate you taking the time to write your response.

**Reply by pderrenger** [reply]
Hey there! 👋

You're absolutely right, and I appreciate your clarification. When we talk about freezing layers during transfer learning, we're generally referring to the early layers of the network. These are kept constant to retain the learned features from the original dataset. The final detection layers, however, are not frozen and are fine-tuned to adapt to the new dataset. This selective freezing allows us to train models more efficiently on new tasks. 

For anyone interested, here's a quick snippet on how to freeze layers in YOLOv8:

```python
# Example: Freeze layers 0-23 for YOLOv8
freeze_layers = range(24)  # layers to freeze
for name, param in model.named_parameters():
    # Freeze specified layers
    if any(f'model.{layer}.' in name for layer in freeze_layers):
        param.requires_grad = False
```

Keep the feedback coming, and happy training! 🚀
